// user-startup/lib/startup/win32.js
'use strict';

var fs = require('fs');
var cp = require('child_process');
var mkdirp = require('mkdirp');
var untildify = require('untildify');

var dir = untildify('~\\AppData\\Roaming\\Microsoft\\Windows\\Start Menu\\Programs\\Startup');

module.exports = {
  dir: dir,
  getFile: getFile,
  add: add,
  create: create,
  remove: remove
};

function getFile(name) {
  return dir + '\\' + name + '.vbs';
}

function add(name, cmd) {
  var args = arguments.length <= 2 || arguments[2] === undefined ? [] : arguments[2];
  var out = arguments[3];

  var file = getFile(name);

  var command = '""' + cmd + '""';

  if (args.length) {
    var escapedArgs = args.map(function (a) {
      return '""' + a + '""';
    }).join(' ');
    command += ' ' + escapedArgs;
  }

  if (out) {
    command += ' > ""' + out + '""';
  }

  var data = 'CreateObject("Wscript.Shell").Run "cmd /c ""' + command + '""", 0, true';

  mkdirp.sync(dir);
  fs.writeFileSync(file, data);
  return file;
}

function create(name, cmd, args, out) {
  var file = add(name, cmd, args, out);

  // Spawn vbscript
  cp.spawn('cmd', ['/c', file], {
    stdio: 'ignore',
    detached: true
  }).unref();
}

function remove(name) {
  var file = getFile(name);
  if (fs.existsSync(file)) fs.unlinkSync(file);
}
